core/model/src/main/java/org/activityinfo/model/util/Holder.java

```diff
@@ -0,0 +1,68 @@
+package org.activityinfo.model.util;
+/*
+ * #%L
+ * ActivityInfo Server
+ * %%
+ * Copyright (C) 2009 - 2013 UNICEF
+ * %%
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, either version 3 of the 
+ * License, or (at your option) any later version.
+ * 
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public 
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/gpl-3.0.html>.
+ * #L%
+ */
+
+/**
+ * @author yuriyz on 11/17/2015.
+ */
+public class Holder<T> {
```

Comment:
It seems you don't like `Holder` class :). `Holder` class keeps reference on `Attachment` object, so we can get/set value from `FormFieldWidget` and pass it to `Uploader` (to set filename, mimetype etc). Advantages of immutable classes are clear but here idea of `Holder` is to have it mutable to share the reference.

https://github.com/bedatadriven/activityinfo/blob/ai-618/server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/ImageUploadFieldWidget.java#L290

Does it look tricky? We can replace `Holder` with `Optional` if add something like

```
class Attachment {
    void set(Attachment attachment) {
          // set all props of attachments
    }
}
```